File: pkg/client/signingidentity.go
/*
Copyright 2020 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package client

import (
	"github.com/golang/protobuf/proto"
	"github.com/hyperledger/fabric-gateway/pkg/hash"
	"github.com/hyperledger/fabric-gateway/pkg/identity"
	"github.com/hyperledger/fabric-protos-go/msp"
)

type signingIdentity struct {
	id   identity.Identity
	sign identity.Sign
	hash hash.Hash
}

func newSigningIdentity(id identity.Identity, sign identity.Sign) *signingIdentity {
	return &signingIdentity{
		id:   id,
		sign: sign,
		hash: hash.SHA256,
	}
}

func (signingID *signingIdentity) Identity() identity.Identity {
	return signingID.id
}

func (signingID *signingIdentity) Hash(message []byte) ([]byte, error) {
	return signingID.hash(message)
}

func (signingID *signingIdentity) Sign(digest []byte) ([]byte, error) {
	return signingID.sign(digest)
}

func (signingID *signingIdentity) Creator() ([]byte, error) {
	serializedIdentity := &msp.SerializedIdentity{
		Mspid:   signingID.id.MspID(),
		IdBytes: signingID.id.Credentials(),
	}
	return proto.Marshal(serializedIdentity)
}

File: node/src/signingidentity.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { msp } from '@hyperledger/fabric-protos';
import { ConnectOptions } from './gateway';
import { Hash } from './hash/hash';
import { sha256 } from './hash/hashes';
import { Identity } from './identity/identity';
import { Signer } from './identity/signer';

export const undefinedSignerMessage = 'No signing implementation';

const undefinedSigner: Signer = () => {
    throw new Error(undefinedSignerMessage);
};

type SigningIdentityOptions = Pick<ConnectOptions, 'identity' | 'signer' | 'hash'>;

export class SigningIdentity {
    readonly #identity: Identity;
    readonly #creator: Uint8Array;
    readonly #hash: Hash;
    readonly #sign: Signer;

    constructor(options: Readonly<SigningIdentityOptions>) {
        this.#identity = {
            mspId: options.identity.mspId,
            credentials: Uint8Array.from(options.identity.credentials),
        };

        const serializedIdentity = new msp.SerializedIdentity();
        serializedIdentity.setMspid(options.identity.mspId);
        serializedIdentity.setIdBytes(options.identity.credentials);
        this.#creator = serializedIdentity.serializeBinary();

        this.#hash = options.hash ?? sha256;
        this.#sign = options.signer ?? undefinedSigner;
    }

    getIdentity(): Identity {
        return {
            mspId: this.#identity.mspId,
            credentials: Uint8Array.from(this.#identity.credentials),
        };
    }

    getCreator(): Uint8Array {
        return Uint8Array.from(this.#creator);
    }

    hash(message: Uint8Array): Uint8Array {
        return this.#hash(message);
    }

    async sign(digest: Uint8Array): Promise<Uint8Array> {
        return this.#sign(digest);
    }
}
